File: src/view-tree-parser.ts
import { Position, Range } from "vscode-languageserver/node";

export interface ParsedComponent {
  name: string;
  range: Range;
  properties: ParsedProperty[];
  startLine: number;
  endLine: number;
}

export interface ParsedProperty {
  name: string;
  range: Range;
  line: number;
  indentLevel: number;
  isBinding: boolean;
  bindingType?: "one-way" | "two-way" | "override" | undefined;
  value?: string;
}

export interface ParsedNode {
  type: "root_class" | "class" | "comp" | "prop" | "sub_prop";
  name: string;
  range: Range;
  line: number;
  indentLevel: number;
}

export interface ParseResult {
  components: ParsedComponent[];
  nodes: ParsedNode[];
  errors: ParseError[];
}

export interface ParseError {
  message: string;
  range: Range;
  severity: "error" | "warning" | "info";
}

export class ViewTreeParser {
  private lines: string[] = [];

  parse(content: string): ParseResult {
    this.lines = content.split("\n");

    const result: ParseResult = {
      components: [],
      nodes: [],
      errors: [],
    };

    let currentComponent: ParsedComponent | null = null;

    for (let lineIndex = 0; lineIndex < this.lines.length; lineIndex++) {
      const line = this.lines[lineIndex];
      if (!line) continue;
      const trimmed = line.trim();

      // Skip empty lines and comments
      if (!trimmed || trimmed.startsWith("//")) {
        continue;
      }

      const indentLevel = this.getIndentLevel(line);

      // Root level component definition
      if (indentLevel === 0 && trimmed.startsWith("$")) {
        // Finish previous component
        if (currentComponent) {
          currentComponent.endLine = lineIndex - 1;
          result.components.push(currentComponent);
        }

        // Start new component
        const firstWord = trimmed.split(/\s+/)[0];
        if (!firstWord) continue;
        const wordRange = this.getWordRange(lineIndex, line.indexOf(firstWord), firstWord);

        currentComponent = {
          name: firstWord,
          range: wordRange,
          properties: [],
          startLine: lineIndex,
          endLine: lineIndex,
        };

        // Add node for root class
        const nodeType = lineIndex === 0 && wordRange.start.character === 1 ? "root_class" : "class";
        result.nodes.push({
          type: nodeType,
          name: firstWord,
          range: wordRange,
          line: lineIndex,
          indentLevel: 0,
        });
      }
      // Property or sub-component
      else if (indentLevel > 0 && currentComponent) {
        const wordMatch = line.match(/^(\s+)([a-zA-Z_$][a-zA-Z0-9_?*]*)/);
        if (wordMatch) {
          const propertyName = wordMatch[2];
          if (!propertyName) continue;
          const propertyStart = line.indexOf(propertyName);
          const wordRange = this.getWordRange(lineIndex, propertyStart, propertyName);

          // Determine if it's a binding
          const isBinding = trimmed.includes("<=") || trimmed.includes("<=>");
          let bindingType: "one-way" | "two-way" | "override" | undefined;
          let value: string | undefined;

          if (isBinding) {
            if (trimmed.includes("<=>")) {
              bindingType = "two-way";
            } else if (trimmed.includes("<=")) {
              bindingType = "one-way";
            }

            // Extract bound property name
            const bindingMatch = trimmed.match(/<=>\s*([a-zA-Z_][a-zA-Z0-9_?*]*)|<=\s*([a-zA-Z_][a-zA-Z0-9_?*]*)/);
            if (bindingMatch) {
              value = bindingMatch[1] || bindingMatch[2];
            }
          } else if (trimmed.includes("^")) {
            bindingType = "override";
          } else {
            // Extract value after property name
            const valueMatch = trimmed.match(/^[a-zA-Z_$][a-zA-Z0-9_?*]*\s+(.+)$/);
            if (valueMatch && valueMatch[1]) {
              value = valueMatch[1].trim();
            }
          }

          const property: ParsedProperty = {
            name: propertyName,
            range: wordRange,
            line: lineIndex,
            indentLevel,
            isBinding,
            bindingType,
            value,
          };

          currentComponent.properties.push(property);

          // Determine node type
          let nodeType: "comp" | "prop" | "sub_prop";
          if (propertyName.startsWith("$")) {
            nodeType = "comp";
          } else if (indentLevel === 1) {
            nodeType = "prop";
          } else {
            nodeType = "sub_prop";
          }

          result.nodes.push({
            type: nodeType,
            name: propertyName,
            range: wordRange,
            line: lineIndex,
            indentLevel,
          });
        }
      }
      // Error: indented line without current component
      else if (indentLevel > 0 && !currentComponent) {
        const errorRange = Range.create(Position.create(lineIndex, 0), Position.create(lineIndex, line.length));
        result.errors.push({
          message: "Property defined outside of component",
          range: errorRange,
          severity: "error",
        });
      }
    }

    // Finish last component
    if (currentComponent) {
      currentComponent.endLine = this.lines.length - 1;
      result.components.push(currentComponent);
    }

    return result;
  }

  getNodeAtPosition(content: string, position: Position): ParsedNode | null {
    const parseResult = this.parse(content);

    for (const node of parseResult.nodes) {
      if (this.isPositionInRange(position, node.range)) {
        return node;
      }
    }

    return null;
  }

  getWordRangeAtPosition(content: string, position: Position): Range | null {
    this.lines = content.split("\n");

    if (position.line >= this.lines.length) {
      return null;
    }

    const line = this.lines[position.line];
    if (!line) {
      return null;
    }
    const character = position.character;

    // Find word boundaries
    let start = character;
    let end = character;

    // Move start backwards to find word start
    while (start > 0 && line[start - 1] && this.isWordCharacter(line[start - 1])) {
      start--;
    }

    // Move end forwards to find word end
    while (end < line.length && line[end] && this.isWordCharacter(line[end])) {
      end++;
    }

    if (start === end) {
      return null;
    }

    return Range.create(Position.create(position.line, start), Position.create(position.line, end));
  }

  getCurrentComponent(content: string, position: Position): string | null {
    this.lines = content.split("\n");

    // Look backwards from current position to find the component
    for (let i = position.line; i >= 0; i--) {
      const line = this.lines[i];
      if (!line) continue;
      const trimmed = line.trim();

      // If line has no indentation and starts with $
      if (!line.startsWith("\t") && !line.startsWith(" ") && trimmed.startsWith("$")) {
        const firstWord = trimmed.split(/\s+/)[0];
        if (firstWord && firstWord.startsWith("$")) {
          return firstWord;
        }
      }
    }

    return null;
  }

  private getIndentLevel(line: string): number {
    let indent = 0;
    for (const char of line) {
      if (char === "\t") {
        indent++;
      } else if (char === " ") {
        indent++; // Could be adjusted for different space-to-tab ratios
      } else {
        break;
      }
    }
    return indent;
  }

  private getWordRange(line: number, start: number, word: string): Range {
    return Range.create(Position.create(line, start), Position.create(line, start + word.length));
  }

  private isPositionInRange(position: Position, range: Range): boolean {
    if (position.line < range.start.line || position.line > range.end.line) {
      return false;
    }

    if (position.line === range.start.line && position.character < range.start.character) {
      return false;
    }

    if (position.line === range.end.line && position.character > range.end.character) {
      return false;
    }

    return true;
  }

  private isWordCharacter(char: string): boolean {
    return /[a-zA-Z0-9_$?*]/.test(char);
  }

  // Helper method to validate view.tree syntax
  validateSyntax(content: string): ParseError[] {
    const parseResult = this.parse(content);
    const errors: ParseError[] = [...parseResult.errors];

    // Add additional validation rules
    for (const component of parseResult.components) {
      // Check for duplicate component names
      const duplicates = parseResult.components.filter((c) => c.name === component.name);
      if (duplicates.length > 1) {
        errors.push({
          message: `Duplicate component name: ${component.name}`,
          range: component.range,
          severity: "warning",
        });
      }

      // Check for invalid property names
      for (const property of component.properties) {
        if (!this.isValidPropertyName(property.name)) {
          errors.push({
            message: `Invalid property name: ${property.name}`,
            range: property.range,
            severity: "error",
          });
        }
      }
    }

    return errors;
  }

  private isValidPropertyName(name: string): boolean {
    // Basic validation - starts with letter or underscore, contains only alphanumeric, underscore, ?, *
    return /^[a-zA-Z_$][a-zA-Z0-9_?*]*$/.test(name);
  }
}

File: lsp/view-tree-parser.go
package main

import (
	"regexp"
	"strings"
)

type ParsedComponent struct {
	Name       string          `json:"name"`
	Range      Range           `json:"range"`
	Properties []ParsedProperty `json:"properties"`
	StartLine  int             `json:"startLine"`
	EndLine    int             `json:"endLine"`
}

type ParsedProperty struct {
	Name        string  `json:"name"`
	Range       Range   `json:"range"`
	Line        int     `json:"line"`
	IndentLevel int     `json:"indentLevel"`
	IsBinding   bool    `json:"isBinding"`
	BindingType string  `json:"bindingType,omitempty"` // "one-way", "two-way", "override"
	Value       string  `json:"value,omitempty"`
}

type ParsedNode struct {
	Type        string `json:"type"` // "root_class", "class", "comp", "prop", "sub_prop"
	Name        string `json:"name"`
	Range       Range  `json:"range"`
	Line        int    `json:"line"`
	IndentLevel int    `json:"indentLevel"`
}

type ParseResult struct {
	Components []ParsedComponent `json:"components"`
	Nodes      []ParsedNode      `json:"nodes"`
	Errors     []ParseError      `json:"errors"`
}

type ParseError struct {
	Message  string             `json:"message"`
	Range    Range              `json:"range"`
	Severity string             `json:"severity"` // "error", "warning", "info"
}

type ViewTreeParser struct {
	lines []string
}

func NewViewTreeParser() *ViewTreeParser {
	return &ViewTreeParser{}
}

func (vtp *ViewTreeParser) Parse(content string) ParseResult {
	vtp.lines = strings.Split(content, "\n")

	result := ParseResult{
		Components: []ParsedComponent{},
		Nodes:      []ParsedNode{},
		Errors:     []ParseError{},
	}

	// Stack to track components by indentation level
	componentStack := make(map[int]*ParsedComponent)
	var rootComponent *ParsedComponent

	for lineIndex, line := range vtp.lines {
		if line == "" {
			continue
		}
		trimmed := strings.TrimSpace(line)

		// Skip empty lines and comments
		if trimmed == "" || strings.HasPrefix(trimmed, "//") {
			continue
		}

		indentLevel := vtp.getIndentLevel(line)

		// Root level component definition
		if indentLevel == 0 && strings.HasPrefix(trimmed, "$") {
			// Finish previous root component
			if rootComponent != nil {
				rootComponent.EndLine = lineIndex - 1
				result.Components = append(result.Components, *rootComponent)
			}

			// Clear component stack for new root component
			componentStack = make(map[int]*ParsedComponent)

			// Start new root component
			fields := strings.Fields(trimmed)
			if len(fields) == 0 {
				continue
			}
			firstWord := fields[0]
			wordRange := vtp.getWordRange(lineIndex, strings.Index(line, firstWord), firstWord)

			rootComponent = &ParsedComponent{
				Name:       firstWord,
				Range:      wordRange,
				Properties: []ParsedProperty{},
				StartLine:  lineIndex,
				EndLine:    lineIndex,
			}
			componentStack[0] = rootComponent

			// Add node for root class
			nodeType := "class"
			if lineIndex == 0 && wordRange.Start.Character == 1 {
				nodeType = "root_class"
			}
			result.Nodes = append(result.Nodes, ParsedNode{
				Type:        nodeType,
				Name:        firstWord,
				Range:       wordRange,
				Line:        lineIndex,
				IndentLevel: 0,
			})
		} else if indentLevel > 0 {
			// Check if this line contains a component reference
			componentRef := vtp.extractComponentReference(line)
			if componentRef != "" {
				// Create new component entry for this indentation level
				wordRange := vtp.getWordRange(lineIndex, strings.Index(line, componentRef), componentRef)
				newComponent := &ParsedComponent{
					Name:       componentRef,
					Range:      wordRange,
					Properties: []ParsedProperty{},
					StartLine:  lineIndex,
					EndLine:    lineIndex,
				}
				componentStack[indentLevel] = newComponent
			}

			// Find the current component for this indentation level
			var currentComponent *ParsedComponent
			for level := indentLevel; level >= 0; level-- {
				if comp, exists := componentStack[level]; exists {
					currentComponent = comp
					break
				}
			}

			if currentComponent != nil {
				// Property or sub-component
				wordMatch := regexp.MustCompile(`^(\s+)([a-zA-Z_$][a-zA-Z0-9_?*]*)`).FindStringSubmatch(line)
				if len(wordMatch) > 2 {
					propertyName := wordMatch[2]
					if propertyName == "" {
						continue
					}
					propertyStart := strings.Index(line, propertyName)
					wordRange := vtp.getWordRange(lineIndex, propertyStart, propertyName)

					// Determine if it's a binding
					isBinding := strings.Contains(trimmed, "<=") || strings.Contains(trimmed, "<=>")
					var bindingType string
					var value string

					if isBinding {
						if strings.Contains(trimmed, "<=>") {
							bindingType = "two-way"
						} else if strings.Contains(trimmed, "<=") {
							bindingType = "one-way"
						}

						// Extract bound property name
						bindingMatch := regexp.MustCompile(`<=>\s*([a-zA-Z_][a-zA-Z0-9_?*]*)|<=\s*([a-zA-Z_][a-zA-Z0-9_?*]*)`).FindStringSubmatch(trimmed)
						if len(bindingMatch) > 1 {
							if bindingMatch[1] != "" {
								value = bindingMatch[1]
							} else if len(bindingMatch) > 2 {
								value = bindingMatch[2]
							}
						}
					} else if strings.Contains(trimmed, "^") {
						bindingType = "override"
					} else {
						// Extract value after property name
						valueMatch := regexp.MustCompile(`^[a-zA-Z_$][a-zA-Z0-9_?*]*\s+(.+)$`).FindStringSubmatch(trimmed)
						if len(valueMatch) > 1 {
							value = strings.TrimSpace(valueMatch[1])
						}
					}

					property := ParsedProperty{
						Name:        propertyName,
						Range:       wordRange,
						Line:        lineIndex,
						IndentLevel: indentLevel,
						IsBinding:   isBinding,
						BindingType: bindingType,
						Value:       value,
					}

					currentComponent.Properties = append(currentComponent.Properties, property)

					// Determine node type
					var nodeType string
					if strings.HasPrefix(propertyName, "$") {
						nodeType = "comp"
					} else if indentLevel == 1 {
						nodeType = "prop"
					} else {
						nodeType = "sub_prop"
					}

					result.Nodes = append(result.Nodes, ParsedNode{
						Type:        nodeType,
						Name:        propertyName,
						Range:       wordRange,
						Line:        lineIndex,
						IndentLevel: indentLevel,
					})
				}
			} else if indentLevel > 0 {
				// Error: indented line without current component
				errorRange := Range{
					Start: Position{Line: lineIndex, Character: 0},
					End:   Position{Line: lineIndex, Character: len(line)},
				}
				result.Errors = append(result.Errors, ParseError{
					Message:  "Property defined outside of component",
					Range:    errorRange,
					Severity: "error",
				})
			}
		}
	}

	// Finish last root component
	if rootComponent != nil {
		rootComponent.EndLine = len(vtp.lines) - 1
		result.Components = append(result.Components, *rootComponent)
	}

	return result
}

func (vtp *ViewTreeParser) GetNodeAtPosition(content string, position Position) *ParsedNode {
	parseResult := vtp.Parse(content)

	for _, node := range parseResult.Nodes {
		if vtp.isPositionInRange(position, node.Range) {
			return &node
		}
	}

	return nil
}

func (vtp *ViewTreeParser) GetWordRangeAtPosition(content string, position Position) *Range {
	vtp.lines = strings.Split(content, "\n")

	if position.Line >= len(vtp.lines) {
		return nil
	}

	line := vtp.lines[position.Line]
	if line == "" {
		return nil
	}
	character := position.Character

	// Find word boundaries
	start := character
	end := character

	// Move start backwards to find word start
	for start > 0 && start-1 < len(line) && vtp.isWordCharacter(rune(line[start-1])) {
		start--
	}

	// Move end forwards to find word end
	for end < len(line) && vtp.isWordCharacter(rune(line[end])) {
		end++
	}

	if start == end {
		return nil
	}

	return &Range{
		Start: Position{Line: position.Line, Character: start},
		End:   Position{Line: position.Line, Character: end},
	}
}

func (vtp *ViewTreeParser) GetCurrentComponent(content string, position Position) string {
	vtp.lines = strings.Split(content, "\n")

	if position.Line >= len(vtp.lines) {
		return ""
	}

	// First, check if current line contains a component reference
	currentLine := vtp.lines[position.Line]
	if componentInLine := vtp.extractComponentFromLine(currentLine); componentInLine != "" {
		return componentInLine
	}

	// Look backwards to find the closest component that owns this position
	currentIndent := vtp.getIndentLevel(currentLine)
	
	for i := position.Line - 1; i >= 0; i-- {
		line := vtp.lines[i]
		if line == "" {
			continue
		}
		
		lineIndent := vtp.getIndentLevel(line)
		
		// If we find a line with less indentation, check if it contains a component
		if lineIndent < currentIndent {
			if componentInLine := vtp.extractComponentFromLine(line); componentInLine != "" {
				return componentInLine
			}
		}
		
		// If line has no indentation and starts with $, it's a root component
		if lineIndent == 0 {
			trimmed := strings.TrimSpace(line)
			if strings.HasPrefix(trimmed, "$") {
				fields := strings.Fields(trimmed)
				if len(fields) > 0 && strings.HasPrefix(fields[0], "$") {
					return fields[0]
				}
			}
		}
	}

	return ""
}

func (vtp *ViewTreeParser) extractComponentFromLine(line string) string {
	// Look for component references like "<= Button $mol_button_major"
	trimmed := strings.TrimSpace(line)
	
	// Check for binding patterns with components
	patterns := []string{
		`<=\s+\w+\s+(\$\w+)`,
		`=>\s+\w+\s+(\$\w+)`,
		`<=>\s+\w+\s+(\$\w+)`,
		`^\s*(\$\w+)`, // Direct component reference
	}
	
	for _, pattern := range patterns {
		re := regexp.MustCompile(pattern)
		if matches := re.FindStringSubmatch(trimmed); len(matches) > 1 {
			return matches[1]
		}
	}
	
	return ""
}

func (vtp *ViewTreeParser) extractComponentReference(line string) string {
	// Extract component reference from binding lines like "<= Button $mol_button_major"
	trimmed := strings.TrimSpace(line)
	
	// Check for component references in bindings
	patterns := []string{
		`<=\s+\w+\s+(\$\w+)`,
		`=>\s+\w+\s+(\$\w+)`,
		`<=>\s+\w+\s+(\$\w+)`,
	}
	
	for _, pattern := range patterns {
		re := regexp.MustCompile(pattern)
		if matches := re.FindStringSubmatch(trimmed); len(matches) > 1 {
			return matches[1]
		}
	}
	
	return ""
}

func (vtp *ViewTreeParser) getIndentLevel(line string) int {
	indent := 0
	for _, char := range line {
		if char == '\t' {
			indent++
		} else {
			break
		}
	}
	return indent
}

func (vtp *ViewTreeParser) getWordRange(line, start int, word string) Range {
	return Range{
		Start: Position{Line: line, Character: start},
		End:   Position{Line: line, Character: start + len(word)},
	}
}

func (vtp *ViewTreeParser) isPositionInRange(position Position, r Range) bool {
	if position.Line < r.Start.Line || position.Line > r.End.Line {
		return false
	}

	if position.Line == r.Start.Line && position.Character < r.Start.Character {
		return false
	}

	if position.Line == r.End.Line && position.Character > r.End.Character {
		return false
	}

	return true
}

func (vtp *ViewTreeParser) isWordCharacter(char rune) bool {
	return (char >= 'a' && char <= 'z') ||
		(char >= 'A' && char <= 'Z') ||
		(char >= '0' && char <= '9') ||
		char == '_' || char == '$' || char == '?' || char == '*'
}

// ValidateSyntax validates view.tree syntax
func (vtp *ViewTreeParser) ValidateSyntax(content string) []ParseError {
	parseResult := vtp.Parse(content)
	errors := make([]ParseError, len(parseResult.Errors))
	copy(errors, parseResult.Errors)

	// Add additional validation rules
	componentNames := make(map[string][]ParsedComponent)
	for _, component := range parseResult.Components {
		componentNames[component.Name] = append(componentNames[component.Name], component)
	}

	// Check for duplicate component names
	for name, components := range componentNames {
		if len(components) > 1 {
			for i := 1; i < len(components); i++ {
				errors = append(errors, ParseError{
					Message:  "Duplicate component name: " + name,
					Range:    components[i].Range,
					Severity: "warning",
				})
			}
		}
	}

	// Check for invalid property names
	for _, component := range parseResult.Components {
		for _, property := range component.Properties {
			if !vtp.isValidPropertyName(property.Name) {
				errors = append(errors, ParseError{
					Message:  "Invalid property name: " + property.Name,
					Range:    property.Range,
					Severity: "error",
				})
			}
		}
	}

	return errors
}

func (vtp *ViewTreeParser) isValidPropertyName(name string) bool {
	// Basic validation - starts with letter or underscore, contains only alphanumeric, underscore, ?, *
	matched, _ := regexp.MatchString(`^[a-zA-Z_$][a-zA-Z0-9_?*]*$`, name)
	return matched
}